lib/elixir/src/elixir_bitstring.erl

```diff
@@ -33,12 +33,22 @@ expand(_BitstrMeta, _Fun, [], Acc, S, E, Alignment, _RequireSize) ->
 expand(BitstrMeta, Fun, [{'::', Meta, [Left, Right]} | T], Acc, S, E, Alignment, RequireSize) ->
   {ELeft, {SL, OriginalS}, EL} = expand_expr(Meta, Left, Fun, S, E),
 
-  MatchOrRequireSize = RequireSize or is_match_size(T, EL),
-  EType = expr_type(ELeft),
-  {ERight, EAlignment, SS, ES} = expand_specs(EType, Meta, Right, SL, OriginalS, EL, MatchOrRequireSize),
-
-  EAcc = concat_or_prepend_bitstring(Meta, ELeft, ERight, Acc, ES, MatchOrRequireSize),
-  expand(BitstrMeta, Fun, T, EAcc, {SS, OriginalS}, ES, alignment(Alignment, EAlignment), RequireSize);
+  case {ELeft, Right} of
+    {{'^', _, [{var, _, _}]}, {Type, _, Atom}} when Type == binary orelse Type == bitstring, is_atom(Atom) ->
```

Comment:
We should not match on Right. Macros have not been expanded nor we have normalized bytes and bits to binary/bitstring. We meed need to move this logic deep inside `expand_specs`.

Reply:
You're right, sorry for that. Will also include tests for these.

My first attempt was deep inside expand_specs, and then went for this simplified version because it was getting unwieldy 😅

Reply:
There is a MatchOrRequireSize argument already. Maybe instead of passing it in, have the expand_specs return ok or size_missing. Then you can decide here to either raise, skip or match on pin?

Reply:
I'm not sure if this is what you had in mind but I did use this argument indeed: [7882446](https://github.com/elixir-lang/elixir/pull/13106/commits/788244602a389e2881596be8b63519fd2578cfbd)